### ISSUE TITLE
[armv7][aot][llvm] App crashes on startup in AndroidTypeManager.RegisterNativeMembers

### ISSUE THREAD

**Issue by mathieubourgeois** [xamarin/xamarin-android#2521]
This issue occurs currently on master and on the d16-0-p2 branch.

- Issue doesn't happen when disabling aot
- On my slightly older version of master (9.1.199.12), disabling llvm works as well. However, I couldn't test on an updated master, since AOT compilation without llvm seems to crash right now.
- Everything works properly if you run arm64-v8a instead.
- The crash occurs inside C# code, but doesn't trigger an exception (as can be seen in the logs below), which leads me to believe that the llvm generated code is wrong. I've taken the liberty of disassembling the function and including it here
[disassembly.txt](https://github.com/xamarin/xamarin-android/files/2669680/disassembly.txt)

### Steps to Reproduce

1. Create a new Single-View App
2. Disable Shared Runtime, Activate AOT+LLVM and build only for armeabi-v7a (You can use [App3.zip](https://github.com/xamarin/xamarin-android/files/2669629/App3.zip) )
3. Build and run (armeabi-v7a or arm64-v8a device, doesn't matter, as long as only armv7 is built)
4. Notice the crash

### Expected Behavior

App actually starts

### Actual Behavior

App crashes

### Version Information

Microsoft Visual Studio Professional 2017 
Version 15.9.3
VisualStudio.15.Release/15.9.3+28307.145
Microsoft .NET Framework
Version 4.7.03062

Installed Version: Professional

Visual C++ 2017   00370-20007-36350-AA982
Microsoft Visual C++ 2017

Application Insights Tools for Visual Studio Package   8.14.11009.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.9.04012.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   15.8.31590
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2017   5.2.60913.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.9.03024.0
Azure App Service Tools v3.0.0

C# Tools   2.10.0-beta2-63501-03+b9fb1610c87cccc8ceb74a770dba261a58e39c4a
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Cookiecutter   15.9.18254.1
Provides tools for finding, instantiating and customizing templates in cookiecutter format.

EclipseAndroidImporter   1.0
Import Android projects from Eclipse.

Extensibility Message Bus   1.1.49 (remotes/origin/d15-8@ee674f3)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

JavaScript UWP Project System   2.0
JavaScript UWP Project System

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.10730.2

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

MLGen Package Extension   1.0
MLGen Package Visual Studio Extension Detailed Info

Mono Debugging for Visual Studio   4.13.11-pre (400cfb1)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Python   15.9.18254.1
Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.

Python - Django support   15.9.18254.1
Provides templates and integration for the Django web framework.

Python - IronPython support   15.9.18254.1
Provides templates and integration for IronPython-based projects.

Python - Profiling support   15.9.18254.1
Profiling support for Python projects.

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   15.1.61810.11040
Microsoft SQL Server Data Tools

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   15.9.20918.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.10.0-beta2-63501-03+b9fb1610c87cccc8ceb74a770dba261a58e39c4a
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual C++ for Cross Platform Mobile Development (Android)   15.0.28107.00
Visual C++ for Cross Platform Mobile Development (Android)

Visual C++ for Cross Platform Mobile Development (iOS)   15.0.28016.00
Visual C++ for Cross Platform Mobile Development (iOS)

Visual F# Tools 10.2 for F# 4.5   15.8.0.0.  Commit Hash: 6e26c5bacc8c4201e962f5bdde0a177f82f88691.
Microsoft Visual F# Tools 10.2 for F# 4.5

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Unity   3.9.0.3
Visual Studio Tools for Unity

Visual Studio Tools for Universal Windows Apps   15.0.28301.59
The Visual Studio Tools for Universal Windows apps allow you to build a single universal app experience that can reach every device running Windows 10: phone, tablet, PC, and more. It includes the Microsoft Windows 10 Software Development Kit.

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   4.12.3.73 (d15-9@20b35b881)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.16.5 (6b9f32067)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   1.1.127 (e975299)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.1.199.12 (HEAD/dd498d2)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: mono/mono/2018-06@b63e5378e38
    Java.Interop: xamarin/java.interop/master@6dd6f16
    LibZipSharp: grendello/LibZipSharp/master@d695149
    LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
    MXE: xamarin/mxe/xamarin@b9cbb535
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.25.2@4ea4c69
    Xamarin.Android Tools: xamarin/xamarin-android-tools/master@9e78d6e

Xamarin.iOS and Xamarin.Mac SDK   12.2.1.11 (5ef73d3)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

### Log File

Logs for the run (this was done using d16-0-p2 branch for test, but it also occurs on master)
[logs.txt](https://github.com/xamarin/xamarin-android/files/2669642/logs.txt)

> VS bug [#782589](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/782589)

**Comment by mathieubourgeois** [comment]
Code of the offending function :

```
		public override void RegisterNativeMembers (JniType jniType, Type type, string methods)
		{
			if (FastRegisterNativeMembers (jniType, type, methods))
				return;

			if (methods == null)
				return;

			string[] members = methods.Split ('\n');
			if (members.Length == 0)
				return;

			JniNativeMethodRegistration[] natives = new JniNativeMethodRegistration [members.Length-1];
			for (int i = 0; i < members.Length; ++i) {
				string method = members [i];
				if (string.IsNullOrEmpty (method))
					continue;
				string[] toks = members [i].Split (new[]{':'}, 4);
				Delegate callback;
				if (toks [2] == "__export__") {
					var mname = toks [0].Substring (2);
					var minfo = type.GetMethods (BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static).Where (m => m.Name == mname && JavaNativeTypeManager.GetJniSignature (m) == toks [1]).FirstOrDefault ();
					if (minfo == null)
						throw new InvalidOperationException (String.Format ("Specified managed method '{0}' was not found. Signature: {1}", mname, toks [1]));
					callback = CreateDynamicCallback (minfo);
				} else {
					GetCallbackHandler connector = (GetCallbackHandler) Delegate.CreateDelegate (typeof (GetCallbackHandler), 
						toks.Length == 4 ? Type.GetType (toks [3], true) : type, toks [2]);
					callback = connector ();
				}
				natives [i] = new JniNativeMethodRegistration (toks [0], toks [1], callback);
			}

			JniEnvironment.Types.RegisterNatives (jniType.PeerReference, natives, natives.Length);
		}
```

From what I could analyse of the assembly, the crash seems to come from

```
					GetCallbackHandler connector = (GetCallbackHandler) Delegate.CreateDelegate (typeof (GetCallbackHandler), 
						toks.Length == 4 ? Type.GetType (toks [3], true) : type, toks [2]);
==>					callback = connector ();
```

The assembly it's related to is

```
   0xc3f83d7c <+808>:   bl      0xc3fef480 <plt_System_Delegate_CreateDelegate_System_Type_System_Type_string>
   0xc3f83d80 <+812>:   cmp     r0, #0
   0xc3f83d84 <+816>:   beq     0xc3f83da4 <Android_Runtime_AndroidTypeManager_RegisterNativeMembers_Java_Interop_JniType_System_Type_string+848>
   0xc3f83d88 <+820>:   ldr     r1, [r0]
   0xc3f83d8c <+824>:   ldr     r2, [sp, #4]
   0xc3f83d90 <+828>:   ldr     r1, [r1]
   0xc3f83d94 <+832>:   ldr     r1, [r1, #8]
   0xc3f83d98 <+836>:   ldr     r1, [r1, #12]
   0xc3f83d9c <+840>:   cmp     r2, r1
   0xc3f83da0 <+844>:   bne     0xc3f83f60 <Android_Runtime_AndroidTypeManager_RegisterNativeMembers_Java_Interop_JniType_System_Type_string+1292> // Invalid cast
   0xc3f83da4 <+848>:   ldr     r1, [r0, #24]
   0xc3f83da8 <+852>:   blx     r1
=> 0xc3f83dac <+856>:   mov     r3, r0
```

Since r0 isn't empty, it means the delegate was created, it was cast properly to `GetCallbackHandler` because it passed the Invalid cast check at +844. My guess now is that the load of the function call to execute might be grabbed from the wrong location (`   0xc3f83da4 <+848>:   ldr     r1, [r0, #24]` where r0 seems to be the GetCallbackHandler object)

**Comment by mathieubourgeois** [comment]
I managed to have a working version of the application by including this in the csproj :

`<AndroidAotAdditionalArguments>llvmopts="-data-layout="p:32:32:32" -O2 -disable-tail-calls"</AndroidAotAdditionalArguments>`

From what I can understand, we need to provide to opt the data layout (http://llvm.org/docs/LangRef.html#data-layout). Usually, this is set directly in the bitcode, but it seems like Mono only sets it in the case of wasm. Otherwise, it looks like llvm makes suppositions in that case about the target data layout which, in that case, seems to mean considering that pointers are 64-bit instead of 32-bit when targeting armeabi-v7a. Sadly, the fix I have is hackish (won't work if you target arm64-v8a at the same time, and I'm not sure what other suppositions opt will make since I just set infos on the pointer sizes, but there are also endianess and alignments constraints)

This seems very much like a Mono issue at this point, I'll go and create an issue in the Mono bug base.

**Comment by pjcollins** [comment]
Thanks for the report and thorough investigation @mathieubourgeois! I see that the follow up issue you filed against mono has been been picked up, however the attempted fix there does not appear to be working for me. I've re-opened that issue and will move this one to the d16-0 milestone as this behavior does appear to be a regression as compared to d15-9.

**Comment by mathieubourgeois** [comment]
I can confirm that I tried the Jenkins artifact versions and they indeed don't work for me either. I haven't tried building it myself though with the fix, so I'm not sure if it's because the Mono cross-compilers have been rebuilt properly or not, because the fix in Mono looks like it should be sufficient. On my end, I did build a custom version where I force add `llvmopts="-data-layout="p:32:32:32" -O2 -disable-tail-calls"` for additional arguments, but only on armeabi-v7a and x86, while waiting for a cleaner fix, which I can confirm works (but is an horrible hack :) )

**Comment by brendanzagaeski** [comment]
**Windows fix published**: The new Xamarin.Android SDK version 9.3 that now includes both parts of the upstream Mono fix for this issue has been published as part of Visual Studio 2019 version 16.1. [Check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) or install the most recent release from <https://visualstudio.microsoft.com/downloads/>.

**macOS fix upcoming**: The fix is not yet published to the **Stable** updater channel in Visual Studio for Mac. For users who have the option of using the **Preview** channel with their projects, the fix is available in the Xamarin.Android SDK version 9.3.0.14 that is currently in the **Preview** updater channel. The promotion to the **Stable** updater channel is upcoming and will align with the promotion of Visual Studio 2019 for Mac version 8.1 to the **Stable** updater channel.

I will leave this issue open for the moment until the macOS fix has been published.

**Comment by brendanzagaeski** [comment]
**macOS fix published**: The new Xamarin.Android SDK version 9.3 that includes both parts of the upstream Mono fix for this issue has now been published the Stable updater channel in Visual Studio for Mac. [Check for the latest updates](https://docs.microsoft.com/visualstudio/mac/update) or install the most recent release from https://visualstudio.microsoft.com/downloads/ to get the fix.

Thanks again for the report and for your investigation work on this issue!
